Custom package repository document expansion and update (#4101)
dart-lang/site-www: src/tools/pub/custom-package-repositories.md

description: "How Dart's package management tool, pub, works with custom package repositories."
---

> The `dart pub` tool has support for third-party package repositories.

Reviewer: ```suggestion
The `dart pub` tool supports third-party package repositories.
```

A package repository is a server that hosts Dart packages
for consumption by the `dart pub` tool.
The default package repository used, [pub.dev]({{site.pub}}), 
is operated by the Dart team to 
facilitate publication of Dart packages for public use.
A package repository is identified by a
_hosted-url_, such as `https://dart-packages.example.com/`.

Sometimes a custom package repository might be useful
for hosting private packages,
including in some of the following scenarios:

1. Sharing internal proprietary packages within an organization.
2. Tight control of dependencies in enterprise environments.
3. Secure environments without public internet access.

> It is also common to use [git-dependencies](/tools/pub/dependencies#git-packages) 

Reviewer: ```suggestion
It's also common to use [git-dependencies](/tools/pub/dependencies#git-packages) 
```

for hosting private packages, however, 
the `dart pub` tool doesn't support resolving versions against a git repository; 
it just fetches a specific revision of the git repository.
Therefore, when many people are collaborating

> it is often preferable to use a private package repository.

Reviewer: ```suggestion
it's often preferable to use a private package repository.
```

  To prevent accidental publication to [pub.dev]({{site.pub}})
  when working on private package 
  it is a good idea to specify this early in the development.
{{site.alert.end}}

To prepare a package for publishing to `https://dart-packages.example.com`,
your `pubspec.yaml` should look minimally like the following:

```yaml
name: example_package
version: 1.0.0
# Ensures the package is published to https://dart-packages.example.com
publish_to: https://dart-packages.example.com
```

To then publish a new version of the package,
use `dart pub publish`:

```terminal
$ dart pub publish
Publishing example_package 1.0.0 to https://dart-packages.example.com
|-- CHANGELOG.md
|-- LICENSE
|-- README.md
|-- lib
|   '-- example_package.dart
'-- pubspec.yaml
...
```

{{site.alert.note}}
  Even if you aren't using a private repository, 
  you can specify `publish_to: none` 
  which stops any accidental publication.
{{site.alert.end}}

### Overriding the default package repository {#default-override}

By default, `dart pub` retrieves dependencies from and publishes packages
to the [pub.dev site]({{site.pub}})
unless the hosted-dependency syntax
is used to specify a custom package repository.

> However, it’s possible to override the default package repository using the

Reviewer: ```suggestion
However, you can override the default package repository using the
```

description: "How Dart's package management tool, pub, works with custom package repositories."
---

The `dart pub` tool has support for third-party package repositories. A package
repository is an HTTP server from which the `pub` client can fetch packages.

> The official package repository is [pub.dev]({{site.pub}}), this repository

Reviewer 2: This line feels awkward.

is by the `dart pub` tool by default. A package repository is identified by a

> _hosted-url_ (e.g. `https://dart-packages.example.com/my-account/`). 

Reviewer: Writing guidelines, we don't use "e.g" or "i.e." replace with "for example" or "such as" as appropriate.

Third-party package servers can be useful for hosting private packages. 
It is also common to use [git-dependencies](/tools/pub/dependencies#git-packages) 
for hosting private packages, however, 

> the `dart pub` tool does not support resolving versions against
> a git-repository it just fetches a specific revision of the git repository.

Reviewer: ```suggestion
the `dart pub` tool doesn't support resolving versions against
a git-repository; it just fetches a specific revision of the git repository.
```
